### PR TITLE
DEV: Add Pre-Commit Hooks To Format Markdown & Notebook Markdown

### PR DESCRIPTION
**Pull request by John-P**
- Add a pre-commit hook (see https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/5abed221d484472e8da436c329b9edfe8d5a7664 and https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/3e6641bab4b50c17ea5286f6ac334d0eef7e3879)
  - Format .md files against GitHub style ([see mdformat-gfm](https://mdformat.readthedocs.io/en/stable/users/plugins.html#parser-extension-plugins)), as these are primarily viewed on GitHub.
  - Format notebook markdown against the [MyST style](https://jupyterbook.org/en/stable/content/myst.html) ([see mdformat-myst](https://mdformat.readthedocs.io/en/stable/users/plugins.html#parser-extension-plugins)). MyST was chosen because it is used by [Jupyter](https://jupyterbook.org/en/stable/content/myst.html) and [Sphinx](https://www.sphinx-doc.org/en/master/usage/markdown.html).
  - Format verbatim blocks with black (e.g. code in ```python markdown blocks via [mdformat-black](https://mdformat.readthedocs.io/en/stable/users/plugins.html#existing-plugins)).
- Move all local repo hooks to the start of the config (run first).
- Migrate Sphinx markdown parsing from recommonmark to MyST due to [deprecation of recommonmark](https://github.com/readthedocs/recommonmark). See https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/7423cfcad641d994559c108aeef28a202dee0f9f. This brings a stronger guarantee that markdown will be consistent between sphinx and notebooks, since they use the same parser now.

## Notes

- **To more easily review only the manually modified files** (i.e. excluding those which have been modified by the hook after it was added) see https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/5abed221d484472e8da436c329b9edfe8d5a7664, https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/3e6641bab4b50c17ea5286f6ac334d0eef7e3879, and https://github.com/TissueImageAnalytics/tiatoolbox/pull/441/commits/7423cfcad641d994559c108aeef28a202dee0f9f. Essentially, any `.md` and `.ipynb` files in the 'Files Changed' section have been modified by the automated hook and not by manual edits, and you may wish to exclude them on your first review pass.
- Pre-commit run log for the run which modified files: https://results.pre-commit.ci/run/github/267705904/1660308570.cLUpP2jGRYSlmKjgwQkDeA
- Although these diffs are large for notebooks now, [this should lead to smaller diffs in the future, e.g. by not using consecutive numbering for enumerated lists](https://mdformat.readthedocs.io/en/stable/users/style.html#) and enforcing consistent JSON formatting for notebooks in pre-commit (this hook uses the python json module to format the code with consistent indentation).
- Other plugins and formatters could be added, e.g. to format bash code in verbatim blocks in addition to python ([see mdformat-shfmt](https://mdformat.readthedocs.io/en/stable/users/plugins.html#existing-plugins)).
- ReviewNB and GitHub will not display the diffs well here due to the indentation changes creating large spans in the diff. If you can, a local tool such as meld or VSCode will display the Diff more easily.
- ~~GitHub Actions appears to be failing on a test for which there are no modified source files. Is this an upstream issue @shaneahmed?~~
- **Read The Docs Build: https://tia-toolbox.readthedocs.io/en/dev-markdown-lint/**

### CODE REVIEW

**Comment by DavidBAEpstein** [comment]
Before I can start this review, I need to have a lot of background explained to me.
Where are the file(s) that test the file(s) being tested. Where is it explained which files will be tested?
The claim in this PR#441 is that 21 files have been changed. Probably because I don’t get on well with ReviewNB, I cannot detect any changes connected with #441 occurrring in any of the example jupyter notebooks. (But I only had the patience to look through a few. I couldn’t see any comparison of “before” with “after” the application of pre-commit hooks.). I think that ReviewNB does not report changes in metadata jsons, although these changes can be important.

ReviewNB’s  get-out from saying anything sensible, is that it doesn’t do large diffs.

It’s not clear to me whether the “Pre-Commit hooks” give suggestions for change, or automatically change the file.

As an example of how mysterious the process is to me, I notice that my PR#435 fails the pre-commit hook associated to flake8. How does one test a jupyter notebook with flake8? I would have expected to see somewhere a comment by flake8 on what it objectss to, but can find none.

The Pre-Commit Hooks approach is a great idea. I wish I had a better understanding of it.
Can I install pre-commit hooks in my own repo, so that my changes are tested before I make a PR?

As I understand it, there are several different flavours of markdown. Is “github markdown” the same as “jupyter notebook markdown”. I have the impression that these are changing, as the Jupyter community tries to agree common syntax with the Readthedocs community

Isn't MyST a parser? We shouldn’t define a language as “What MyST accepts”, since then programming errors could be labelled as “features”.

David

**Comment by John-P** [comment]
@DavidBAEpstein

1. I suggest you look at the links in the notes section above and read [the pre-commit documentation](https://pre-commit.com/). It will answer most of your questions e.g. "Where is it explained which files will be tested" you should see the `.pre-commit-config.yml` config `types_or` key for the new notebook markdown hook.
2. In addition, I would suggest taking a look at markdown specifications e.g. [CommonMark](https://commonmark.org/), [MyST Markdown](https://jupyterbook.org/en/stable/content/myst.html), and [GitHub Flavored Markdown](https://github.github.com/gfm/).
3. I suggest not using ReviewNB for this PR. It will not be that helpful. The diffs are unfortunately too large for GitHub to display. They do however, display well locally e.g. in VSCode.

**Comment by John-P** [comment]
@DavidBAEpstein Please see also the updated PR description for more detailed information.

**Review by DavidBAEpstein — COMMENTED** [review]
I reviewed changes only in files for which ReviewNB gives a reasonable report. Finding both necessary and unnecessary white space was good, and, presumably, the corresponding changes were made without direct human editing. I have some comments about the substitution of the HTML tag "&mdash;" by a glyph. First of all, as can be seen from the ReviewNB display of examples/README.md, this change is noticed by ReviewNB in some cases, but not in all cases. My poor opinion of ReviewNB is correspondlingly reinforced. Secondly, I wondered why this change was thought necessary. The glyph seemed, on at least one occasion, too short for an emdash. (I checked the utf8 codes, and they were indeed different. Then I decided that it didn't matter, since I may be unique in being bothered by such very minor points.) Because I have confidence in the author of this PR, I approve it notwithstanding the fact that I haven't been able to do as thorough a job as I would have liked.

**Comment by DavidBAEpstein** [comment]
PR441
I don't know how to react to this message (see screen shot)

**Comment by shaneahmed** [comment]
> PR441
> 
> I don't know how to react to this message (see screen shot)

Hi @DavidBAEpstein, you do not need to worry about this. I will do this once it is ready to merge.

File: README.md [hunk 1]
```diff
@@ -50,26 +50,25 @@ This package is for those interested in digital pathology: including graduate st
...
 
-
 ### Examples Taster
 
 1. [Click here](https://github.com/TissueImageAnalytics/tiatoolbox/tree/develop/examples) for [jupyter notebooks](https://jupyter.org/), hosted on the web, with demos of `tiatoolbox`. All necessary resources to run the notebooks are remotely provided, so you don't need to have Python installed on your computer.
-2. Click on a filename with suffix `.ipynb` and the notebook will open in your browser.
-3. Click on one of the two blue checkboxes in your browser window labelled either **Open in Colab** or **Open in Kaggle**: [colab](https://colab.research.google.com/notebooks/intro.ipynb#) and [kaggle](https://www.kaggle.com/) are websites providing free-of-charge platforms for running jupyter notebooks.
-4. Operate the notebook in your browser, editing, inserting or deleting cells as desired.
-5. Changes you make to the notebook will last no longer than your colab or kaggle session.
+1. Click on a filename with suffix `.ipynb` and the notebook will open in your browser.
```
Comment (shaneahmed) [inline README.md:68]: It has rendered correctly but I am not sure if this should be `1.`

Comment (John-P) [reply README.md:68]: See https://mdformat.readthedocs.io/en/stable/users/style.html#, the 1. numbering is used to reduce diff size when making edits. The number before the period is NOT used when determining the number when the document is rendered. This is purely done from ordering. Hence, using 1. for all items reduces edit diff size when changing lists.

**Review by shaneahmed — APPROVED** [review]
Thanks @John-P 

Looks good. Except that benchmarks/annotation_store.ipynb needs to be added to Example Gallery.
Please do it in a separate PR.

**Pull request merged**